README.md
# html-json2canvas

基于类 html 语法的 JSON 配置，使用 Canvas 绘制网页界面；
在线示例： <https://json2canvas.netlify.app/>

## 安装

You can install it with npm or `<script>`:

```bash
npm install html-json2canvas
```

```html
<script src="./dist/json2canvas.umd.min.js"></script>
```

## 用法

```js
const json2canvas = require("json2canvas");
const options = {
  ratio: 2,
  config: {
    type: "div",
    x: 0,
    y: 0,
    css: {
      width: 200,
      radius: 10,
    },
    children: [
      {
        type: "div",
        css: {
          padding: [8, 16, 8, 16],
          backgroundColor: "#268dff",
          opacity: 0.7,
          textAlign: "center",
        },
        children: [
          {
            type: "text",
            text: "我是标题",
            css: {
              lineHeight: 20,
              display: "inline-block",
              color: "#fff",
              fontSize: 14,
            },
          },
        ],
      },
      {
        type: "div",
        css: {
          padding: [6, 8, 6, 8],
          backgroundColor: "#FFF",
          opacity: 0.7,
        },
        children: [
          {
            type: "div",
            css: {
              padding: [5, 0, 5, 0],
            },
            children: [
              {
                type: "text",
                text: "我是Label：",
                css: {
                  width: 80,
                  color: "#1f1f1f",
                  fontSize: 13,
                  lineHeight: 20,
                  textAlign: "right",
                  display: "inline-block",
                },
              },
              {
                type: "text",
                text: "我是Value",
                css: {
                  color: "#1f1f1f",
                  fontSize: 13,
                  lineHeight: 20,
                  display: "inline-block",
                },
              },
            ],
          },
        ],
      },
    ],
  },
};

(async () => {
  const canvas = await json2canvas(options);

  document.body.appendChild(canvas);
})();
```

## 参数配置

| name   | required | default  | description               |
| ------ | :------: | :------: | ------------------------- |
| config |    ✔️    |    {}    | canvas elements config    |
| width  |          | 自动计算 | canvas width              |
| height |          | 自动计算 | canvas height             |
| ratio  |          | 自动计算 | canvas pixel ratio        |
| render |          |  'html'  | canvas render: html, node |

### Elements JSON Schema

#### div

```js
{
  type: 'div',
  x: 0, // 仅限根元素，子元素自动计算
  y: 0, // 仅限根元素，子元素自动计算
  css: {
    width: 200,
    radius: 10,
    padding: [8, 16, 8, 16],
    backgroundColor: '#268dff',
    opacity: 0.7,
    textAlign: 'center'
  },
  children: []
}
```

#### image

```js
{
  type: 'image',
  url: '', // 支持跨域图片
  css: {
    width: 200,
    height: 100,
    radius: 10,
    opacity: 0.7
  }
}
```

#### text

```js
{
  type: 'text',
  text: '',
  css: {
    color: '#1f1f1f',
    fontSize: 14,
    lineHeight: 20,
    width: 80,
    textAlign: 'right',
    display: 'inline-block'
  }
}
```

#### wrapText

```js
{
  type: 'wrapText',
  text: '',
  css: {
    color: '#1f1f1f',
    fontSize: 14,
    width: 200, // 到宽度边界时换行
    height: 80, // 超过高度的文字会被截掉
    lineHeight: 20,
    lineClamp: 2, // 文字超过两行结尾追加"..."
    textAlign: 'right',
    display: 'inline-block'
  }
}
```

#### custom function

```js
{
  type: function ({ x, y, url, css }) {
    const { ctx, canvas, ratio } = this
  },
  x: 0,
  y: 0,
  url: '',
  // text: '',
  css: {}
}
```
> 自定义函数类型：可使用canvas API自由绘制一些基本类型不能满足的或者复杂的元素

## Example

启动本地服务： http://localhost:3000

在线示例: <https://json2canvas.netlify.app/>

## License

json2canvas is released under the MIT license.


CHANGELOG.md
# [1.1.0-1](https://github.com/wangqh/json2canvas/compare/1.1.0-0...1.1.0-1) (2022-02-11)


### Bug Fixes

* 两端对齐时适配安卓 ios 下的文字居中问题 ([9f72f44](https://github.com/wangqh/json2canvas/commit/9f72f44c2ed402aff92fb85f09d6254fb645a9d3))



# [1.1.0-0](https://github.com/wangqh/json2canvas/compare/1.0.5...1.1.0-0) (2022-02-09)


### Bug Fixes

* 解决htmlCanvasExtends引用路径错误 ([efafa6c](https://github.com/wangqh/json2canvas/commit/efafa6ce26eefa0946cff83e1f7aa961275feb92))
* 重复声明变量报错 ([e64b793](https://github.com/wangqh/json2canvas/commit/e64b79318878ffbdc1a4b5669da364dac81bac5e))


### Features

* 添加letterSpacingText方法，实现两端对齐效果 ([5233953](https://github.com/wangqh/json2canvas/commit/52339535ac05b0b7c9b34aefa7ba8dc889baa85e))
* 增加支持css的textAlign: justify两端对齐配置， canvas context原型扩展letterSpacingText方法 ([4051063](https://github.com/wangqh/json2canvas/commit/4051063d399ab7aa991b247e2ecc124e241ca2a4))
* 增加支持css的textAlign: justify两端对齐配置， canvas context原型扩展letterSpacingText方法 ([9ebd69b](https://github.com/wangqh/json2canvas/commit/9ebd69ba8627742c931b9b416747e74f26804860))
* 增加支持css的textAlign: justify两端对齐配置， canvas context原型扩展letterSpacingText方法 ([8725a87](https://github.com/wangqh/json2canvas/commit/8725a87ae8cfcf2ccbbee270de33b1e5bc85da59))



## [1.0.5](https://github.com/wangqh/json2canvas/compare/1.0.4...1.0.5) (2021-10-21)


### Bug Fixes

* wrapText类型的输入文字undefined时报异常问题 ([c723232](https://github.com/wangqh/json2canvas/commit/c7232324382745dd68650b4c7d42d288fe506224))



## [1.0.4](https://github.com/wangqh/json2canvas/compare/1.0.3...1.0.4) (2021-10-18)


### Bug Fixes

* wrapText类型不设置高度时，动态计算高度错误，且不支持lineClamp ([dad0f75](https://github.com/wangqh/json2canvas/commit/dad0f75cea5fe76948e5d742718bf99cde8641d7))



## [1.0.3](https://github.com/wangqh/json2canvas/compare/1.0.2...1.0.3) (2021-10-15)



## [1.0.2](https://github.com/wangqh/json2canvas/compare/1.0.1...1.0.2) (2021-09-30)


### Features

* 增加绘制图片类型Demo ([b426eaf](https://github.com/wangqh/json2canvas/commit/b426eafaddfeca15fc14a2bf9661a6f876cb399b))



## [1.0.1](https://github.com/wangqh/json2canvas/compare/1.0.0...1.0.1) (2021-09-27)


### Bug Fixes

* 透明度为0时无效问题 ([97595e4](https://github.com/wangqh/json2canvas/commit/97595e40be3754628d90f45541bbb72286e99041))



# 1.0.0 (2021-09-26)


### Bug Fixes

* 删除注释的代码 ([f374f29](https://github.com/wangqh/json2canvas/commit/f374f2971a93be461f1a668a7949754d5c6b381c))


### Performance Improvements

* 允许设定宽高 ([66eafef](https://github.com/wangqh/json2canvas/commit/66eafefe7ee0d53526c63b2019f1f6537542b474))





demos/simple/index.js
const options = {
  ratio: 2,
  config: {
    type: "div",
    x: 0,
    y: 0,
    css: {
      width: 200,
      radius: 10,
    },
    children: [
      {
        type: "div",
        css: {
          padding: [8, 16, 8, 16],
          backgroundColor: "#268dff",
          opacity: 0.7,
          textAlign: "center",
        },
        children: [
          {
            type: "text",
            text: "我是标题",
            css: {
              lineHeight: 20,
              display: "inline-block",
              color: "#fff",
              fontSize: 14,
            },
          },
        ],
      },
      {
        type: "div",
        css: {
          padding: [6, 8, 6, 8],
          backgroundColor: "#FFF",
          opacity: 0.7,
        },
        children: [
          {
            type: "div",
            css: {
              padding: [5, 0, 5, 0],
            },
            children: [
              {
                type: "text",
                text: "我是Label：",
                css: {
                  width: 80,
                  color: "#1f1f1f",
                  fontSize: 13,
                  lineHeight: 20,
                  textAlign: "right",
                  display: "inline-block",
                },
              },
              {
                type: "text",
                text: "我是Value",
                css: {
                  color: "#1f1f1f",
                  fontSize: 13,
                  lineHeight: 20,
                  display: "inline-block",
                },
              },
            ],
          },
        ],
      },
    ],
  },
};

(async () => {
  const canvas = await json2canvas(options);
  document.body.appendChild(canvas);
})();